### PR TITLE
Documentation for Fetch Metadata and Coop/Coep Suppport

### PR DESCRIPTION
**Pull request by eozmen410**

### CODE REVIEW

File: wicket-user-guide/src/main/asciidoc/security/security_5.adoc [hunk 1]
```diff
@@ -2,7 +2,40 @@
 
 _CryptoMapper_ helps preventing CSRF attacks by making the urls impossible to be guessed by an attacker but still there is some theoretical chance this to happen.
 
-To further help against this kind of vulnerability Wicket provides _CsrfPreventionRequestCycleListener_ - a _IRequestCycleListener_ that forbids requests made from a different origin. By default only actions are forbidden, i.e. a request coming from different origin cannot execute _Link.onClick()_ or submit forms (_Form.onSubmit()_). Any request to render pages are still allowed so Wicket pages could be easily embedded in other applications.
+To further help against this kind of vulnerability Wicket provides _ResourceIsolationRequestCycleListener_ - a _IRequestCycleListener_ that uses __IResourceIsolationPolicy__ objects to decide whether to allow or reject a request. By default only actions are checked, i.e. a request coming from different origin cannot execute _Link.onClick()_ or submit forms (_Form.onSubmit()_). Any request to render pages are still allowed so Wicket pages could be easily embedded in other applications.
```
Comment (salcho) [inline wicket-user-guide/src/main/asciidoc/security/security_5.adoc]:
```suggestion
To further help against this kind of vulnerability Wicket provides _ResourceIsolationRequestCycleListener_ - a _IRequestCycleListener_ that uses __IResourceIsolationPolicy__ objects to decide whether to allow or reject cross-origin requests. By default only actions are checked, i.e. a cross-origin request cannot execute _Link.onClick()_ or submit forms (_Form.onSubmit()_). Any request to render pages are still allowed so Wicket pages could be easily embedded in other applications.
```

File: wicket-user-guide/src/main/asciidoc/security/security_5.adoc [hunk 2]
```diff
@@ -2,7 +2,40 @@
...
+ protected void init() {
+  super.init();
+  getRequestCycleListeners().add(new ResourseIsolationRequestCycleListener());
+  // ...
+ }
+----
+
+_ResourceIsolationRequestCycleListener_ is highly configurable. It allows to add exempted paths that will not be checked with the __addExemptedPath__ method. It can be configured with multiple _ResourceIsolationPolicy_ objects to be checked in order.
+
+An __IResourceIsolationPolicy__ returns a __ResourceIsolationOutcome__ after processing a request, which can be one of 3 values (__ALLOWED__, __DISALLOWED__, __UNKNOWN__). The __ResourceIsolationRequestCycleListener__ checks the __IResourceIsolationPolicy__ objects in order and uses the first outcome that is not __UNKNOWN__ to trigger the appropriate action. If all return __UNKNOWN__ __unknownOutcomeAction__ is applied. The actions can be configured through the listener.
```
Comment (salcho) [inline wicket-user-guide/src/main/asciidoc/security/security_5.adoc:20]: what if an outcome returns ALLOWED? we keep iterating over the remaining policies, right? If so, this should read: 'the first outcome that is not __UNKNOWN__ or __ALLOWED__', I believe

File: wicket-user-guide/src/main/asciidoc/security/security_5.adoc [hunk 3]
```diff
@@ -19,3 +52,4 @@ _CsrfPreventionRequestCycleListener_ is highly configurable. It allows to define
 
 _CsrfPreventionRequestCycleListener_ is not an alternative to _CryptoMapper_! Both of them could be used separately or in tandem to prevent CSRF attacks depending on the application requirements.
 
+
```
Comment (salcho) [inline wicket-user-guide/src/main/asciidoc/security/security_5.adoc]: let's remove this extra whitespace if it isn't needed, please

**Review by salcho — CHANGES_REQUESTED** [review]
thank you J!

File: wicket-user-guide/src/main/asciidoc/security/security_5.adoc (hunk 2 above)
Comment (eozmen410) [reply wicket-user-guide/src/main/asciidoc/security/security_5.adoc:20]: I think if the outcome is `ALLOWED` we [stop iterating over the remaining policies and return](https://github.com/apache/wicket/blob/0f42d3314f1200a3b88f76e435910ef389af6b58/wicket-core/src/main/java/org/apache/wicket/protocol/http/ResourceIsolationRequestCycleListener.java#L336-L339)

Comment (salcho) [reply wicket-user-guide/src/main/asciidoc/security/security_5.adoc:20]: right! not our brightest moment 🤦‍♂️

**Review by salcho — APPROVED** [review]
💯

**Comment by eozmen410** [comment]
I added the COOP and COEP explanations! I named the file `security_cross_origin_isolation.adoc` for now, but I'm assuming we will need to change it something like `security_7.adoc` and modify the numbers for the other files (change the current `securit_7.adoc` to `security_8.adoc` etc.) to keep convention (since I added it after csp in `single.adoc`)

File: wicket-user-guide/src/main/asciidoc/security/security_cross_origin_isolation.adoc [hunk 4]
```diff
@@ -0,0 +1,25 @@
+A major consequence of transient execution attacks for the web is the risk of exposing private authenticated data to untrusted sites via microarchitectural side channels. Some web APIs increase the risk of side-channel attacks like https://meltdownattack.com/[Spectre]. To mitigate that risk, browsers offer an opt-in-based isolated environment called cross-origin isolated using Cross-Origin Opener Policy and Cross-Origin Embedder Policy.
```
Comment (salcho) [inline wicket-user-guide/src/main/asciidoc/security/security_cross_origin_isolation.adoc]:
```suggestion
A major consequence of transient execution attacks for the web is the risk of exposing private authenticated data to untrusted sites via microarchitectural side channels. Some web APIs increase the risk of side-channel attacks like https://meltdownattack.com/[Spectre]. To mitigate that risk, browsers offer two opt-in mechanisms to effectively isolate resources from third-party origins through Cross-Origin Opener Policy and Cross-Origin Embedder Policy.
```

File: wicket-user-guide/src/main/asciidoc/security/security_cross_origin_isolation.adoc [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy[Cross-Origin Embedder Policy (COEP)] prevents a document from loading any cross-origin resources which don't explicitly grant the document permission to be loaded.
+
+COOP and COEP are independent mechanisms that can be enabled, tested and deployed separately. While enabling one doesn’t require developers to enable the other, when set together COOP and COEP allows developers to use powerful features (such as __SharedArrayBuffer__, __performance.measureMemory()__ and the JS Self-Profiling API) securely, without worrying about side channel attacks like Spectre. Further reading on https://docs.google.com/document/d/1zDlfvfTJ_9e8Jdc8ehuV4zMEu9ySMCiTGMS9y0GU92k/edit#bookmark=id.uo6kivyh0ge2[COOP/COEP] and https://web.dev/why-coop-coep/[why you need cross-origin isolation].
```
Comment (salcho) [inline wicket-user-guide/src/main/asciidoc/security/security_cross_origin_isolation.adoc]:
```suggestion
COOP and COEP are independent mechanisms that can be enabled, tested and deployed separately. While enabling one doesn’t require developers to enable the other, when set together COOP and COEP allows developers to use powerful features (such as __SharedArrayBuffer__, __performance.measureMemory()__ and the JS Self-Profiling API) securely, while still mitigating information leaks and side channel attacks like Spectre. Read more about COOP/COEP on https://docs.google.com/document/d/1zDlfvfTJ_9e8Jdc8ehuV4zMEu9ySMCiTGMS9y0GU92k/edit#bookmark=id.uo6kivyh0ge2[COOP/COEP] and https://web.dev/why-coop-coep/[why you need cross-origin isolation].
```

File: wicket-user-guide/src/main/asciidoc/security/security_cross_origin_isolation.adoc [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+Wicket provides highly configurable support for COOP and COEP. Users can specify the mode for each policy. For COOP, the __CoopMode__ represents the three valid header values and disabling COOP enitrely. __CoopMode__ can be one of the four values __SAME_ORIGIN__, __SAME_ORIGIN_ALLOW_POPUPS__, __UNSAFE_NONE__ and __DISABLED__. For COEP the __CoepMode__ represents whether the policy will be __Report-Only__ (__REPORTING__ mode), enforcing (__ENFORCING__ mode) or disabled entirely (__DISABLED__). For both COOP and COEP, exempted paths can be specified, the response headers will not be set for these paths. Exemptions can include paths that use features that need to hold JavaScript references to windows opened through the __window.open__ function.
```
Comment (salcho) [inline wicket-user-guide/src/main/asciidoc/security/security_cross_origin_isolation.adoc]: let's list the three header values here for COOP like you did for COEP please

**Review by salcho — CHANGES_REQUESTED** [review]
As you say, chances are we'll have to change the name of the files. I wouldn't worry too much about making the file numbers match their section (that is, instead of naming this security_7 (because it's the seventh section), I'd name it security_n+1 where n is the latest security_* file. In general, if you choose to leave the file named, I'd go for resource_isolation and not cross_origin_isolation.

File: wicket-user-guide/src/main/asciidoc/security/security_10.adoc [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+Wicket provides highly configurable support for COOP and COEP. Users can specify the mode for each policy. For COOP, the __CoopMode__ represents the three valid header values __same-origin__ (__SAME_ORIGIN__ mode), __same-origin-allow-popups__ (__SAME_ORIGIN_ALLOW_POPUPS__ mode), __unsafe-none__ (__UNSAFE_NONE__ mode) and disabling COOP enitrely (__DISABLED__). For COEP the __CoepMode__ represents whether the policy will be __Report-Only__ (__REPORTING__ mode), enforcing (__ENFORCING__ mode) or disabled entirely (__DISABLED__). For both COOP and COEP, exempted paths can be specified, the response headers will not be set for these paths. Exemptions can include paths that use features that need to hold JavaScript references to windows opened through the __window.open__ function.
```
Comment (salcho) [inline wicket-user-guide/src/main/asciidoc/security/security_10.adoc]: Let's be explicit about the fact that the set of exempted paths expects relative paths with trailing slashes, please. See related comments on the doc PR for Struts.

**Review by salcho — CHANGES_REQUESTED** [review]

File: wicket-user-guide/src/main/asciidoc/security/security_10.adoc (hunk 7 above)
Comment (eozmen410) [reply wicket-user-guide/src/main/asciidoc/security/security_10.adoc]: good point!

**Review by salcho — APPROVED** [review]
Thanks Jen! LGTM! 🥇

**Pull request merged**